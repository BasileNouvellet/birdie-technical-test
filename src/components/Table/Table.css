.table {
    width: 80%;
    border-radius: 10px;
    background-color: white;
    box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23);
    overflow: hidden;
}

.rows-container {
    overflow: auto;
    max-height: 65vh;
}

.rows-container::-webkit-scrollbar-track {
    margin: 2px 0 0 0;
}

.rows-container::-webkit-scrollbar {
    width: 8px;
}

.rows-container::-webkit-scrollbar-thumb {
    border-radius: 10px;
    background-color: #9a9a9a;
}

.footer {
    height: 3em;
    width: 15%;
    margin: 1em auto;
    display: flex;
    justify-content: center;
    align-items: center;
    color: white;
    font-weight: bold;
    border: none;
    border-radius: 10px;
    outline: none;
    background-color: #6c61c9;
}
