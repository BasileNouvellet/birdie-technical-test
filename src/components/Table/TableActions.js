// @flow

export const TABLE_ACTIONS_TYPES = {
    changeVariable: 'TABLE_ACTIONS_TYPES/changeVariable',
    fetchData: 'TABLE_ACTIONS_TYPES/fetchData',
    dataFetched: 'TABLE_ACTIONS_TYPES/dataFetched',
};

export function changeVariableAction(variable) {
    return {
        type: TABLE_ACTIONS_TYPES.changeVariable,
        payload: {
            variable,
        },
    };
}

export function fetchDataAction(variable) {
    return {
        type: TABLE_ACTIONS_TYPES.fetchData,
        payload: {
            variable,
        },
    };
}

export function dataFetchedAction(data) {
    return {
        type: TABLE_ACTIONS_TYPES.dataFetched,
        payload: {
            data,
        },
    };
}
