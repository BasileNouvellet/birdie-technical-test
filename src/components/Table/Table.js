// @flow

import * as React from 'react';
import { connect } from 'react-redux';

import Row from './subComponents/Row';
import Spinner from './subComponents/Spinner';
import ErrorMessage from '../ErrorMessage';

import './Table.css';

import type {
  TableStateType,
  TableStateVariableType,
  TableStateDataType,
  TableStateErrorType,
} from './index';

type MappedStatePropsType = {|
  variable: TableStateVariableType,
  data: TableStateDataType,
  error: TableStateErrorType,
  loading: TableStateLoadingType,
|};
type OwnPropsType = {||};
type PropsType = MappedStatePropsType & OwnPropsType;

type StateType = {|
  shortList: boolean,
|};

// export unconnected component for test purposes with Jest
export class TableUnconnected extends React.Component<PropsType, StateType> {
  state: StateType = {
    shortList: true,
  };

  componentDidUpdate(prevProps: PropsType) {
    const { variable } = this.props;
    if (prevProps.variable !== variable) this.resetShortList(); // variable has changed
  }

  getCapitalizedVariable(): ?string {
    const { variable } = this.props;

    if (!variable) return null;

    return variable
      .split(' ')
      .map(word => word.charAt(0)
        .toUpperCase() + word.slice(1))
      .join(' ');
  }

  onFooterClick = (): void => {
    this.setState((state: StateType) => ({
      shortList: !state.shortList,
    }));
  };

  getSlicedRows(): TableStateDataType {
    const { data } = this.props;
    const { shortList } = this.state;

    const shouldBeSliced = shortList && data.length >= 100;

    return shouldBeSliced
      ? data.slice(0, 100)
      : data;
  }

  resetShortList() {
    const { shortList } = this.state;
    if (!shortList) this.setState({ shortList: true }); // reset shortlist to true
  }

  renderColumnsNames(): ?React.Element<Row> {
    const capitalizedVariable = this.getCapitalizedVariable();

    if (!capitalizedVariable) return null;

    return (
      <Row
        index="#"
        variable={capitalizedVariable}
        count="Count"
        averageAge="Average Age"
        isTitle
      />
    );
  }

  renderRows(): Array<React.Element<Row>> {
    const { variable } = this.props;

    return this.getSlicedRows()
      .map((row, index) => (
        <Row
          key={`${row[variable]}${row.count}${row.averageAge}`}
          index={index}
          variable={row[variable]}
          count={row.count}
          averageAge={row.average_age}
        />
      ));
  }

  renderFooter(): ?React.Element<'button'> {
    const { variable, data } = this.props;
    const { shortList } = this.state;

    if (
      !variable // no variable selected
      || !data.length // or no data to show
      || data.length <= 100 // or less than 100 rows
    ) {
      return null;
    }

    const footerTitle = shortList
      ? `Show all rows (${data.length - 100} more)`
      : 'Hide rows';

    return (
      <button
        className="footer"
        type="button"
        onClick={this.onFooterClick}
      >
        {footerTitle}
      </button>
    );
  }

  // ------------------------------------------- Render ------------------------------------------
  render(): React.Element<'div'> {
    const { error, loading } = this.props;
    if (error.data) return <ErrorMessage />;

    return (
      <div className="table">
        {this.renderColumnsNames()}

        {
          loading ? <Spinner /> : (
            <div className="rows-container">
              {this.renderRows()}
              {this.renderFooter()}
            </div>
          )
        }
      </div>
    );
  }
}

const mapStateToProps = (state: TableStateType): MappedStatePropsType => ({
  variable: state.variable,
  data: state.data,
  error: state.error,
  loading: state.loading,
});

export default connect(
  mapStateToProps,
  null,
)(TableUnconnected);
