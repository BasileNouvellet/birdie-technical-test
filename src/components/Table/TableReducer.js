// @flow

import { TABLE_ACTIONS_TYPES } from './TableActions';

const TableReducer = (state = {
    columns: [],
    variable: null,
    data: [],
}, action) => {
    switch (action.type) {
        case TABLE_ACTIONS_TYPES.setVariable:
            const { variable } = action.payload;

            return {
                ...state,
                variable,
            };
        case TABLE_ACTIONS_TYPES.dataFetched:
            const { data } = action.payload;

            return {
                ...state,
                data,
            };
            case TABLE_ACTIONS_TYPES.columnsFetched:
            const { columns } = action.payload;

            return {
                ...state,
                columns,
            };
        default:
            return state;
    }
};

export default TableReducer;
