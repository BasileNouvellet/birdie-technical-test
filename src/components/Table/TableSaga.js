// @flow

import {
  call, put, all, takeLatest,
} from 'redux-saga/effects';

import { types } from './TableActions';
import { TableActions } from './index';
import type { FetchDataActionType } from './index';

function handleErrors(res) {
  if (!res.ok) {
    return res.json().then((jsonErr) => { throw jsonErr; });
  }

  return res;
}

function fetchApi(url) {
  return fetch(url)
    // since fetch will succeed even if http status code is wrong,
    // we need to check the response ourselves
    // fetch will fail only when there is no network connection or access denied
    .then(handleErrors)
    .then(res => res.json())
    .then(json => ({ response: json })) // return an object with response key
    .catch(error => ({ error })); // return an object with error key
}

function* fetchDataSaga(action: FetchDataActionType): * {
  const { variable } = action.payload;

  if (variable) {
    const url = `/database/data?variable=${variable}`;
    const { response, error } = yield call(fetchApi, url); // either response or error is undefined

    if (response) { // success
      yield put(TableActions.setDataAction(response));
    } else { // error
      yield put(TableActions.fetchDataFailedAction(error, variable));
    }
  }
}

function* fetchColumnsSaga(): * {
  const { response, error } = yield call(fetchApi, '/database/columns');

  if (response && response.length) { // success
    const columns = response.map(rawColumn => rawColumn.Field); // take map only columns names
    yield put(TableActions.setColumnsAction(columns));
  } else { // error
    yield put(TableActions.fetchColumnsFailedAction(error));
  }
}

function* TableSaga(): * {
  yield all([
    takeLatest(types.FETCH_COLUMNS, fetchColumnsSaga),
    takeLatest(types.FETCH_DATA, fetchDataSaga),
  ]);
}

export default TableSaga;
