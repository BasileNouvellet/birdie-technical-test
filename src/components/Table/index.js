import Table from './Table';
import * as TableActions from './TableActions';
import TableReducer from './TableReducer';
import TableSaga from './TableSaga';

export {
  Table as default,
  TableActions,
  TableReducer,
  TableSaga,
};
