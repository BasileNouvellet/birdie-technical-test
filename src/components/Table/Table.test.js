// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

import { TableUnconnected as Table } from './Table';

describe('Table component', () => {
  let wrapper;

  const reduxStoreProps = {
    // map state to props
    variable: 'column2',
    data: [
      {
        column1: 'Male',
        count: 100,
        average_age: 50,
      },
      {
        column2: 'Female',
        count: 1000,
        average_age: 100,
      },
    ],
    error: {
      column: false,
      data: false,
    },
    loading: false,
  };

  beforeEach(() => {
    wrapper = shallow(
      <Table
        variable={reduxStoreProps.variable}
        data={reduxStoreProps.data}
        error={reduxStoreProps.error}
        loading={reduxStoreProps.loading}
      />,
    );
  });

  it('should shallow render without crashing', () => {}); // done in beforeEach()

  it('should match snapshot', () => { expect(wrapper).toMatchSnapshot(); });
});
