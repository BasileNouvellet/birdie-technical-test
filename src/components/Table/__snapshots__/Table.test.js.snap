// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TableUnconnected
    data={
      Array [
        Object {
          "average_age": 50,
          "column1": "Male",
          "count": 100,
        },
        Object {
          "average_age": 100,
          "column2": "Female",
          "count": 1000,
        },
      ]
    }
    error={
      Object {
        "column": false,
        "data": false,
      }
    }
    loading={false}
    variable="column2"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Row
          averageAge="Average Age"
          count="Count"
          index="#"
          isTitle={true}
          variable="Column2"
        />,
        <div
          className="rows-container"
        >
          <Row
            averageAge={50}
            count={100}
            index={0}
            isTitle={false}
          />
          <Row
            averageAge={100}
            count={1000}
            index={1}
            isTitle={false}
            variable="Female"
          />
        </div>,
      ],
      "className": "table",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "averageAge": "Average Age",
          "count": "Count",
          "index": "#",
          "isTitle": true,
          "variable": "Column2",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [
              <Row
                averageAge={50}
                count={100}
                index={0}
                isTitle={false}
              />,
              <Row
                averageAge={100}
                count={1000}
                index={1}
                isTitle={false}
                variable="Female"
              />,
            ],
            null,
          ],
          "className": "rows-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "undefined100undefined",
            "nodeType": "function",
            "props": Object {
              "averageAge": 50,
              "count": 100,
              "index": 0,
              "isTitle": false,
              "variable": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "Female1000undefined",
            "nodeType": "function",
            "props": Object {
              "averageAge": 100,
              "count": 1000,
              "index": 1,
              "isTitle": false,
              "variable": "Female",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Row
            averageAge="Average Age"
            count="Count"
            index="#"
            isTitle={true}
            variable="Column2"
          />,
          <div
            className="rows-container"
          >
            <Row
              averageAge={50}
              count={100}
              index={0}
              isTitle={false}
            />
            <Row
              averageAge={100}
              count={1000}
              index={1}
              isTitle={false}
              variable="Female"
            />
          </div>,
        ],
        "className": "table",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "averageAge": "Average Age",
            "count": "Count",
            "index": "#",
            "isTitle": true,
            "variable": "Column2",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [
                <Row
                  averageAge={50}
                  count={100}
                  index={0}
                  isTitle={false}
                />,
                <Row
                  averageAge={100}
                  count={1000}
                  index={1}
                  isTitle={false}
                  variable="Female"
                />,
              ],
              null,
            ],
            "className": "rows-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "undefined100undefined",
              "nodeType": "function",
              "props": Object {
                "averageAge": 50,
                "count": 100,
                "index": 0,
                "isTitle": false,
                "variable": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "Female1000undefined",
              "nodeType": "function",
              "props": Object {
                "averageAge": 100,
                "count": 1000,
                "index": 1,
                "isTitle": false,
                "variable": "Female",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            null,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
